lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs

```diff
@@ -0,0 +1,396 @@
+//! Asynchronous interface to WebGL GLSL shader program compilation.
+//!
+//! # Performance
+//!
+//! In order to maximize parallelism and avoid blocking, follows
+//! [best practices](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/WebGL_best_practices).
+//!
+//! In particular, these items from that document are implemented here:
+//! - "Compile Shaders and Link Programs in parallel"
+//! - "Prefer KHR_parallel_shader_compile"
+//! - "Don't check shader compile status unless linking fails"
+//!
+//! # Context loss
+//!
+//! The compiler handles [context loss](https://www.khronos.org/webgl/wiki/HandlingContextLost) and
+//! does not report compilation errors when the context is not available.
+
+use crate::prelude::*;
+use crate::system::gpu::context::native::traits::*;
+use crate::system::web::traits::*;
+
+use crate::animation;
+use crate::display::ToGlEnum;
+use crate::system::gpu::context::native;
+use crate::system::gpu::shader;
+use crate::system::gpu::shader::Fragment;
+use crate::system::gpu::shader::Shader;
+use crate::system::gpu::shader::Vertex;
+use crate::system::web;
+
+use web_sys::WebGl2RenderingContext;
+
+
+
+// =================
+// === Constants ===
+// =================
+
+/// We do not want the framerate to drop below this value when compiling shaders. Whenever we
+/// discover that it drops below this threshold, no more shaders will be scheduled for compilation
+/// or linking in this frame. It does not mean, however, that the framerate will not be lower. The
+/// last scheduled shader can take a lot of time to be rendered, causing the FPS to drop
+/// significantly.
```

Comment:
Does this mean if for whatever reason the application is, e.g., limited to 30FPS, no shaders will ever be compiled? That would be bad.

Reply:
No, at least one shader will be processed per frame. Added explanation there.